Ios 2961 refactoring blockchain sdk parth of wallet manager factory

skibinalexander opened the pull request:
- Инкапсулировал логику сборки кошелька в конкретный файл Assembly. В файле структура данных подписаннаю на абстракцию сборки кошелька.
- Blockchain по аналогии с другими параметрами знает своего сборщика, который в свою очередь вызывает фабрика. Безопасно для добавление нового .case сразу отсветит что нужно сделать Assembly для блокчейна
- Сделал структуру для конфигурации провайдеров. Различные блокчейны по одинаковому конфигурировали провайдеры. В дальнейшем поможет дальше отрефачить инициализацию фабрику провайдера. (ex. BlockBookUtxo) 

- Закидываю пока на ревью для обсуждения, волью сам если что

skibinalexander committed:
IOS-2961 Add assembly for all blockchain types

skibinalexander committed:
IOS-2961 Completed making get providers for blockchains

skibinalexander committed:
IOS-2961 Did complete refactor wallet assembly

skibinalexander committed:
IOS-2961 Add refactor private method Blockchair

skibinalexander committed:
IOS-2961 Add @available(iOS 13.0, *)

Balashov152 commented on BlockchainSdk/Common/WalletAssembly.swift:
```diff
@@ -0,0 +1,49 @@
+    let networkConfig: NetworkProviderConfiguration
+}
+
+enum BlockBookProviderType {
+    case NowNodes
+    case GetBlock
+}
```
Лучше в отдельный файл
и case с маленькой )

Balashov152 replied:
Да я вообще раскидал на три файла

BlockchainAssemblyInput, BlockBookProviderType, WalletAssemblyProtocol

skibinalexander replied:
BlockchainAssemblyInput, WalletAssemblyProtocol

Точно стоит, просто в одном файлике по логике они в сильной связке

Balashov152 commented on BlockchainSdk/Common/ProviderAssembly.swift:
```diff
@@ -0,0 +1,63 @@
+    
+    // MARK: - Private Implementation
+    
+    private func makeBlockchairNetworkProviders(for endpoint: BlockchairEndpoint, configuration: NetworkProviderConfiguration, apiKeys: [String]) -> [AnyBitcoinNetworkProvider] {
+        let apiKeys: [String?] = [nil] + apiKeys
+        
+        let providers = apiKeys.map {
+            BlockchairNetworkProvider(endpoint: endpoint, apiKey: $0, configuration: configuration)
+                .eraseToAnyBitcoinNetworkProvider()
+        }
+        
+        return providers
+    }
```
можно метод убрать, и перенести в makeBlockchairNetworkProviders
Он же один раз только вызывается

Balashov152 commented on BlockchainSdk/Common/WalletAssembly.swift:
```diff
@@ -0,0 +1,49 @@
+    /// Blockchain assembly method
+    /// - Parameter input: Input data factory
+    /// - Returns: Blockchain result
+    static func make(with input: BlockchainAssemblyInput) throws -> AssemblyWallet
```
👍

Balashov152 commented on BlockchainSdk/Common/WalletAssembly/DashWalletAssembly.swift:
```diff
@@ -0,0 +1,41 @@
+                bip: .bip44
+            )
+            
+            // TODO: Add CryptoAPIs for testnet
```
Удали плиз, моя TODO, уже не актуально )

ba1ashov commented:
прикольней стало )

megakoko commented on BlockchainSdk/Common/Blockchain.swift:
```diff
@@ -100,6 +100,15 @@ public enum Blockchain: Equatable, Hashable {
         }
     }
     
+    public var shelly: Bool {
```
shelley

megakoko commented on BlockchainSdk/Common/Blockchain.swift:
```diff
@@ -911,3 +920,47 @@ extension Blockchain {
+@available(iOS 13.0, *)
+extension Blockchain {
+    
+    var assembly: WalletAssemblyProtocol.Type {
+        switch self {
+        case .bitcoin:
+            return BitcoinWalletAssembly.self
+        case .litecoin:
+            return LitecoinWalletAssembly.self
```
Может лучше сделать метод make статическим и тут создавать нужный объект?

megakoko replied:
Имею в виду наоборот переделать метод со статического

skibinalexander replied:
Честно сказать хотел сделать именно так, чтобы не делать инстансы, ибо не нужны при статике, достаточно только типа, ну чтобы разночтений не было, здесь вызываем только тип ассембли, дальше фабрика сделает все за тебя

tureck1y replied:
имхо чем меньше статики, тем лучше, зачем ей сидеть в памяти место занимать постоянно

megakoko commented on BlockchainSdk/Common/ProviderAssembly.swift:
```diff
@@ -0,0 +1,63 @@
+        }
+    }
+    
+    func makeInfoNetworkProvider(with input: BlockchainAssemblyInput) -> BlockchainInfoNetworkProvider {
```
makeBlockchainInfoNetworkProvider

Это название сервиса

skibinalexander replied:
> makeBlockchainInfoNetworkProvider
> 
> Это название сервиса

Тут не понял ес честно, что имеешь в виду?

Balashov152 replied:
метод переменововать как я понял )

megakoko replied:
Да, метод надо переименовать в makeBlockchainInfoNetworkProvider.

BlockchainInfo это полное название сервиса

megakoko commented on BlockchainSdk/Common/ProviderAssembly.swift:
```diff
@@ -0,0 +1,63 @@
+        )
+    }
+    
+    func makeBlockchairNetworkProviders(endpoint: BlockchairEndpoint, with input: BlockchainAssemblyInput) -> [AnyBitcoinNetworkProvider] {
+        return makeBlockchairNetworkProviders(
+            for: endpoint,
+            configuration: input.networkConfig,
+            apiKeys: input.blockchainConfig.blockchairApiKeys
+        )
+    }
+    
+    // MARK: - Private Implementation
+    
+    private func makeBlockchairNetworkProviders(for endpoint: BlockchairEndpoint, configuration: NetworkProviderConfiguration, apiKeys: [String]) -> [AnyBitcoinNetworkProvider] {
```
Надо бы единообразие

    func makeBlockchairNetworkProviders(endpoint: ...)
    func makeBlockchairNetworkProviders(for endpoint: ...)

megakoko replied:
Ну или объединить, да

skibinalexander replied:
> Ну или объединить, да

Тут объединю ага

megakoko commented on BlockchainSdk/Common/WalletAssembly.swift:
```diff
@@ -0,0 +1,49 @@
+import Combine
+import TangemSdk
+
+typealias AssemblyWallet = WalletManager
```
Это еще надо?

skibinalexander replied:
> Это еще надо?

Тут все-таки предлагаю оставить "AssemblyWallet" потому что , могут появляться другие протоколы, и чтобы в каждом ассембли потом не делать подпись на дополнительный протокол, он сразу подсветит нужные методы для conform протокола, что выглядит достаточно удобным и безопасным.

megakoko commented on BlockchainSdk/Common/WalletManagerFactory.swift:
```diff
@@ -67,379 +70,32 @@ public class WalletManagerFactory {
-                bip: .bip44
+        return try blockchain.assembly.make(
+            with: .init(
+                blockchain: blockchain,
+                blockchainConfig: config,
+                publicKey: publicKey,
+                pairPublicKey: pairPublicKey,
+                wallet: Wallet(
+                    blockchain: blockchain,
+                    addresses: blockchain.makeAddresses(from: publicKey.blockchainKey, with: pairPublicKey),
+                    publicKey: publicKey
+                ),
```
Если передается Wallet то можно не передавать blockchain/publicKey

megakoko commented on BlockchainSdk/Common/WalletAssembly/KusumaWalletAssembly.swift:
```diff
@@ -0,0 +1,26 @@
+import Foundation
+import TangemSdk
+
+struct KusumaWalletAssembly: WalletAssemblyProtocol {
+    
+    static func make(with input: BlockchainAssemblyInput) throws -> AssemblyWallet {
+        let network: PolkadotNetwork = input.blockchain.isTestnet ? .westend : .polkadot
+        
+        return PolkadotWalletManager(network: network, wallet: input.wallet).then {
+            let providers = network.urls.map { url in
+                PolkadotJsonRpcProvider(url: url, configuration: input.networkConfig)
+            }
+            $0.networkService = PolkadotNetworkService(providers: providers, network: network)
+            $0.txBuilder = PolkadotTransactionBuilder(walletPublicKey: input.wallet.publicKey.blockchainKey, network: network)
+        }
+    }
+    
+}
```
У кусамы нет тестнета и для мейннета network неправильный.

Есть .polkadot — это мейннет полькадот, .westend — тестнет полькадот, и .kusama.

Все работает через одну ж... sonrpc, надо сделать через один метод. Можно тоже в ProviderAssembly затолкать.

megakoko replied:
Или как-то объединить. Можно сделать SubstrateWalletAssembly который принимал бы network.

megakoko replied:
Substrate — это название протокола, который их всех объединяет

megakoko replied:
Наподобие эфира

skibinalexander replied:
Понял, спасибо проверю еще разок

skibinalexander committed:
IOS-2961 Merge branch 'develop' into IOS-2961-Refactoring-blockchain-sdk-parth-of-WalletManagerFactory

skibinalexander committed:
IOS-2961 Fix merge request comments

skibinalexander committed:
IOS-2961 Add Substrate wallet assembly for union Polkadot and Kusama

megakoko commented:
Еще не смотрел, но надо подлить девелоп и убедиться, что все последние изменения в URLах смержились корректно.

https://github.com/tangem/blockchain-sdk-swift/pull/223/files
https://github.com/tangem/blockchain-sdk-swift/pull/220/files

Плюс этот надо будет подлить когда будет готов

https://github.com/tangem/blockchain-sdk-swift/pull/225

megakoko commented:
Этот тоже:

https://github.com/tangem/blockchain-sdk-swift/pull/226

skibinalexander committed:
IOS-2961 Merge branch 'develop' into IOS-2961-Refactoring-blockchain-sdk-parth-of-WalletManagerFactory

# Conflicts:
#	BlockchainSdk.xcodeproj/project.pbxproj
#	BlockchainSdk/Common/Blockchain.swift
#	BlockchainSdk/Common/WalletManagerFactory.swift

skibinalexander committed:
IOS-2961 Fix comment WalletAssembly named

tureck1y commented on BlockchainSdk/Common/Blockchain.swift:
```diff
@@ -100,6 +100,15 @@ public enum Blockchain: Equatable, Hashable {
         }
     }
     
+    public var shelley: Bool {
```
Некрасиво, что эта штука только для карданы нужна, и то все что не Shelley = legacy, а будет болтаться на весь блокчейн

skibinalexander replied:
> Некрасиво, что эта штука только для карданы нужна, и то все что не Shelley = legacy, а будет болтаться на весь блокчейн

Согласен , подумаю как бы нам обойти это

tureck1y commented on BlockchainSdk/Common/WalletAssembly.swift:
```diff
@@ -0,0 +1,44 @@
+    /// Blockchain assembly method
+    /// - Parameter input: Input data factory
+    /// - Returns: Blockchain result
+    static func make(with input: BlockchainAssemblyInput) throws -> WalletAssembly
```
А зачем static? И без него имхо хорошо будет)

tureck1y commented on BlockchainSdk/Common/WalletAssembly.swift:
```diff
@@ -0,0 +1,44 @@
+import Foundation
+import TangemSdk
+
+typealias WalletAssembly = WalletManager
```
- Я бы предложил WalletManager оставить как он есть, чтобы не путаться
- WalletAssemblyProtocol  переименовать в WalletManagerAssembly, потому что Wallet это Wallet, а WalletManager это WalletManager)
- BlockchainAssemblyInput переименовать в WalletManagerAssemblyInput или WalletManagerAssembly.Input
- ProviderAssembly переименовать в NetworkProviderAssembly

skibinalexander replied:
Тут все ок сделаю

tureck1y reviewed (commented):
Прикольно получилось, надо не забыть сюда тоггл перетащить

skibinalexander committed:
IOS-2961 Merge branch 'develop' into IOS-2961-Refactoring-blockchain-sdk-parth-of-WalletManagerFactory

# Conflicts:
#	BlockchainSdk.xcodeproj/project.pbxproj
#	BlockchainSdk/Common/WalletManagerFactory.swift

skibinalexander committed:
IOS-2961 Rename structures by review comment

skibinalexander committed:
IOS-2961 Fix review comments

skibinalexander committed:
IOS-2961 Refactor bitcoin api for WalletAssembly

skibinalexander committed:
IOS-2961 Update Wallet assemblies for api changes

skibinalexander committed:
IOS-2961 Update wallet manager factory

skibinalexander commented:
Все замечания учел просьба еще раз посмотреть

tureck1y commented on BlockchainSdk/Common/ProviderAssembly.swift:
```diff
@@ -0,0 +1,51 @@
+//
+//  ProviderAssembly.swift
```
Название файла осталось старое

Balashov152 commented on BlockchainSdk/Common/ProviderAssembly.swift:
```diff
@@ -0,0 +1,51 @@
+    }
+    
+    func makeBlockchairNetworkProviders(endpoint: BlockchairEndpoint, with input: WalletManagerAssemblyInput) -> [AnyBitcoinNetworkProvider] {
+        let apiKeys: [String?] = [nil] + input.blockchainConfig.blockchairApiKeys
```
Это что бы без ключа провайдер создать? Такое нам нужно?)

Balashov152 commented on BlockchainSdk/Common/WalletAssembly/EthereumWalletAssembly.swift:
```diff
@@ -0,0 +1,71 @@
+            blockcypherProvider = nil
+        }
+        
+        // TODO: Move this generation into assembly.
```
можно удалить наверно, `already moved`

Andoran90 commented on BlockchainSdk/Common/ProviderAssembly.swift:
```diff
@@ -0,0 +1,51 @@
+        }
+    }
+    
+    func makeInfoNetworkProvider(with input: WalletManagerAssemblyInput) -> BlockchainInfoNetworkProvider {
```
```suggestion
    func makeBlockchainInfoNetworkProvider(with input: WalletManagerAssemblyInput) -> BlockchainInfoNetworkProvider {
```

Andoran90 commented on BlockchainSdk/Common/WalletAssembly.swift:
```diff
@@ -0,0 +1,42 @@
+    
+    // MARK: - Wallet Assembly
+    
+    /// Blockchain assembly method
+    /// - Parameter input: Input data factory
+    /// - Returns: Blockchain result
+    func make(with input: WalletManagerAssemblyInput) throws -> WalletManager
```
Давай подправим документацию, а то она не несет особо информации.
```suggestion
    /// Function that creates WalletManager according to input data
    /// - Parameter input: Input that contains information about blockchain, SdkConfig, network settings
    /// - Returns: WalletManager for specified blockchain
    func make(with input: WalletManagerAssemblyInput) throws -> WalletManager
```

Andoran90 commented on BlockchainSdk/Common/WalletAssembly.swift:
```diff
@@ -0,0 +1,42 @@
+protocol WalletManagerAssembly {
+    
+    /// Assembly to access any providers
+    var providerAssembly: NetworkProviderAssembly { get }
```
```suggestion
    var networkProviderAssembly: NetworkProviderAssembly { get }
```

Andoran90 commented on BlockchainSdk/Common/WalletAssembly/BitcoinWalletAssembly.swift:
```diff
@@ -0,0 +1,60 @@
+
+struct BitcoinWalletAssembly: WalletManagerAssembly {
+    
+    // MARK: - Implementation
```
можно убрать

tureck1y commented on BlockchainSdk/Common/WalletAssembly/SubstrateWalletAssembly.swift:
```diff
@@ -0,0 +1,28 @@
+import Foundation
+import TangemSdk
+
+struct SubstrateWalletAssembly: WalletManagerAssembly {
```
KusumaWalletAssembly ?

skibinalexander replied:
Обговаривали @megakoko решили соединить в один assemlby

megakoko replied:
Давай сделаем PolkadotWalletAssembly чтобы не выбивалось

megakoko commented on BlockchainSdk/Common/WalletAssembly/SubstrateWalletAssembly.swift:
```diff
@@ -0,0 +1,28 @@
+import Foundation
+import TangemSdk
+
+struct SubstrateWalletAssembly: WalletManagerAssembly {
```
Substrate — это название протокола, под которым ходят полька, кусама и т.д.

WalletManager, networkprovider и т.д. в названии содержат Polkadot, я так делал чтобы проще было находить в проекте.

skibinalexander committed:
IOS-2961 Fix review comments for any assembly

skibinalexander committed:
IOS-2961 Refactor optimism and eth wallet assembly

tureck1y commented on BlockchainSdk/Common/WalletAssembly/EthereumWalletAssembly.swift:
```diff
@@ -0,0 +1,71 @@
+            manager = OptimismWalletManager(wallet: input.wallet, rpcURL: endpoints[0])
+        } else {
+            manager = EthereumWalletManager(wallet: input.wallet)
+        }
```
Только если отделять, то аккуратно, без копипасты и так, что если что-то поменяем для evm, то Optimism нельзя было бы упустить. Все же минорно отличается

skibinalexander replied:
> Только если отделять, то аккуратно, без копипасты и так, что если что-то поменяем для evm, то Optimism нельзя было бы упустить. Все же минорно отличается

Сань, а как такой вариант, вынес rpc в network assembly provider, чтобы не ошибиться потом если что, и вынес создание RPC провайдера вроде логично выглядит

Andoran90 commented on BlockchainSdk/Common/WalletAssembly/EthereumWalletAssembly.swift:
```diff
@@ -0,0 +1,53 @@
+        
+        var transactionHistoryProvider: TransactionHistoryProvider?
+        
+        if input.blockchain.canLoadTransactionHistory {
+            // This should be decided by each assembly
+            transactionHistoryProvider = BlockscoutNetworkProvider(configuration: .init(credentials: input.blockchainConfig.blockscoutCredentials))
+        }
```
я имел ввиду в тудушке, что надо создание `BlockscoutNetworkProvider` перенести в `NetworkProviderAssembly`
Может даже стоит перенести этот ифчик в `NetworkProviderAssembly`? @tureck1y

tureck1y replied:
Я тоже думал, даже вот тоггл например, но не уверен, что абстрактная фабрика сетевого слоя должна знать о таком)

Andoran90 replied:
это да, согласен :)

skibinalexander committed:
IOS-2961 Update make union eth or optimism implementation

Balashov152 commented on BlockchainSdk/Common/Blockchain.swift:
```diff
@@ -935,3 +935,49 @@ extension Blockchain {
+        case .ethereum, .ethereumClassic, .rsk, .bsc, .polygon, .avalanche, .fantom, .arbitrum, .gnosis, .ethereumPoW, .optimism, .ethereumFair, .saltPay:
+            return EthereumWalletAssembly()
+        case .optimism:
+            return OptimismWalletAssembly()
```
вообще варнинг должен быть)

Andoran90 commented on BlockchainSdk/Common/NetworkProviderAssembly.swift:
```diff
@@ -0,0 +1,66 @@
+        return canLoad ? BlockscoutNetworkProvider(configuration: .init(credentials: input.blockchainConfig.blockscoutCredentials)) : nil
+    }
+    
+    func makeJsonRpcEndpoints(with input: WalletManagerAssemblyInput) -> [URL] {
```
думаю не очень хорошо, что возвращает не NetworkProvider, а массив URL. Было бы лучше, если бы возвращал `EthereumJsonRpcProvider`. Все же ассембли провайдеров

skibinalexander replied:
> думаю не очень хорошо, что возвращает не NetworkProvider, а массив URL. Было бы лучше, если бы возвращал `EthereumJsonRpcProvider`. Все же ассембли провайдеров

Понял, в целом сначала думал так сделать, смутила специфичность eth

skibinalexander committed:
IOS-2961 Update review comments for receive eth network providers
